project_generator/workspace.py

```diff
@@ -12,68 +12,94 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import os
 import yaml
 import logging
 
-from .project import Project
+from .project import Project, ProjectWorkspace
 from .settings import ProjectSettings
 from .tool import ToolsSupported
 from .targets import Targets
+from .util import flatten, uniqify, load_yaml_records
 
 
-class Workspace:
+class PgenWorkspace:
 
     """a collections of projects from a single projects.yaml file"""
 
     def __init__(self, projects_file='projects.yaml', project_root='.'):
-        # load projects file
-
-        with open(projects_file, 'rt') as f:
-            self.projects_dict = yaml.load(f)
+        # either it's file or a dictionary. If dictionary , proceed otherwise load yaml
+        if type(projects_file) is not dict:
+            try:
+                with open(projects_file, 'rt') as f:
+                    self.projects_dict = yaml.load(f)
+            except IOError:
+               raise IOError("The main pgen projects file %s doesn't exist." % projects_file)
+        else:
+            self.projects_dict = projects_file
 
         self.settings = ProjectSettings()
-
         if 'settings' in self.projects_dict:
             self.settings.update(self.projects_dict['settings'])
 
         # so that we can test things independently of eachother
+        self.workspaces = {}
         self.projects = {}
 
+        # We support grouping of projects which is a workspace
+        #
+        # [projects.yaml]
+        #   projects:
+        #       project_group_1:
+        #           project_1:
+        #               -a
+        #               -b
+        #               -c
+        #           project_2:
+        #               -d
+        #               -e
+        #               -f
+        #       project_3:
+        #           -g
+        #           -h
+        # extension - workspaces
+
         if 'projects' in self.projects_dict:
-            for name,records in self.projects_dict['projects'].items():
-                if "common" in records:
-                    self.projects[name] = Project(name, records, self)
+            for name, records in self.projects_dict['projects'].items():
```

Comment:
They should be added although I am still not 100 convinced about settings work group of projects.. But we need them